python.d/linux_power_supply.chart.py

```diff
@@ -0,0 +1,159 @@
+GET_CHART = {
+    'capacity': _get_capacity_chart,
+    'charge': _get_charge_chart,
+    'energy': _get_energy_chart,
+    'voltage': _get_voltage_chart
+}
+
+# This opens the specified file and returns the value in it or None if
+# the file doesn't exist.
+def _get_sysfs_value(filepath):
+    try:
+        with open(filepath, 'r') as datasource:
+            return int(datasource.read())
+    except (OSError, IOError):
+        return None
+
+# Certain attributes (*_empty and *_empty_design) are implicitly zero
+# if not provided by the driver.  This function is used to fetch those
+# attributes.
+def _get_sysfs_value_or_zero(filepath):
+    result = _get_sysfs_value(filepath)
+    if result is None:
+        return 0
+    return result
+
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.definitions = dict()
+        self.order = list()
+        self.attrlist = list()
+        self.supply = self.configuration.get('supply', None)
+        if self.supply is not None:
+            self.syspath = '/sys/class/power_supply/{0}'.format(self.supply)
+        self.types = self.configuration.get('charts', 'energy charge capacity voltage').split()
+
+    def check(self):
+        if self.supply is None:
+            self.error('No power supply specified for monitoring.')
+            return False
+        if not len(self.types) > 0:
+            self.error('No attributes requested for monitoring.')
+            return False
+        if os.access(self.syspath, os.R_OK):
+            chartset = set(GET_CHART.keys()).intersection(set(self.types))
+            if len(chartset) > 0:
+                charts = dict()
+                attrlist = list()
+                for item in chartset:
+                    chart, attrs = GET_CHART[item](self.syspath)
+                    if chart is not None:
+                        charts.update(chart)
+                        attrlist.extend(attrs)
+                if len(charts) == 0:
+                    self.error('No charts can be created.')
+                    return False
+                self.definitions.update(charts)
+                self.order.extend(sorted(charts.keys()))
+                self.attrlist.extend(attrlist)
+            else:
+                self.error('No valid attributes requested for monitoring.')
+                return False
+        else:
+            self.error('Unable to access {0}'.format(self.syspath))
+            return False
+        return True
+
+    def _get_data(self):
+        data = dict()
+        for attr in self.attrlist:
+            attrpath = os.path.join(self.syspath, attr)
+            if attr.endswith('_min') or attr.endswith('_min_design') or \
+               attr.endswith('_empty') or attr.endswith('_empty_design'):
+                data[attr] = _get_sysfs_value_or_zero(attrpath)
```

Comment:
use `any`
```python
any([
	COND1,
	COND2,
	COND3,
])
```

Reply:
I actually did not know about the `any()` function, and had completely forgotten that `str.startswith()` could take a tuple.